src/transformers/cache_utils.py

```diff
@@ -203,10 +403,49 @@ def get_mask_sizes(self, cache_position: torch.Tensor, layer_idx: int) -> tuple[
         The masks are then prepared according to the given lengths (kv_length, kv_offset) and patterns (i.e. sliding_window, chunk_size),
         for each layer.
         """
-        query_length = cache_position.shape[0]
-        past_seen_tokens = self.get_seq_length()
-        kv_length = query_length + past_seen_tokens
-        return kv_length, 0
+        if isinstance(self.layers[layer_idx], SlidingWindowLayer):
+            query_length = cache_position.shape[0]
+            first_cache_position = cache_position[0]
+
+            local_mask_kv_offset = torch.clamp(first_cache_position - self.config.sliding_window + 1, min=0)
+            # This is not general (see HybridChunkedCache for the whole general case), but it's what the cache returns
+            local_mask_kv_length = max(query_length, self.config.sliding_window)
+            return local_mask_kv_length, local_mask_kv_offset
+
+        full_mask_kv_offset = 0
+        if isinstance(self.layers[layer_idx], StaticLayer):
+            full_mask_kv_length = self.get_max_cache_shape()
+            return full_mask_kv_length, full_mask_kv_offset
+        else:
+            query_length = cache_position.shape[0]
+            past_seen_tokens = self.get_seq_length()
+            kv_length = query_length + past_seen_tokens
+            return kv_length, full_mask_kv_offset
```

Comment:
the goal of having layer cache is to only call `self.layer[layer_idx].get_mask_sizes()` instead of having these if elses